ddtrace/contrib/django/patch.py

```diff
@@ -1,17 +1,513 @@
+        patch_restframework(pin.tracer)
+    except Exception:
+        log.exception('Error patching rest_framework')
+
+    # Instrument Databases
+    try:
+        for conn in django.db.connections.all:
+            patch_conn(pin.tracer, conn)
+    except Exception:
+        log.exception('error instrumenting Django database connections')
+
+    # Instrument Caches
+    try:
+        cache_backends = set([cache['BACKEND'] for cache in django.conf.settings.CACHES.values()])
+        for cache_module in cache_backends:
+            for method in ['get', 'set', 'add', 'delete', 'incr', 'decr',
+                           'get_many', 'set_many', 'delete_many', ]:
+                wrap(cache_module, method, traced_cache(django))
+    except Exception:
+        log.exception('error instrumenting Django caches')
+
+    return ret
+
+
+def traced_func(django, name, resource=None):
+    """Returns a function to trace Django functions."""
+    def wrapped(django, pin, func, instance, args, kwargs):
+        with pin.tracer.trace(name, resource=resource):
+            return func(*args, **kwargs)
+    return with_traced_module(wrapped)(django)
+
+
+@with_traced_module
+def traced_load_middleware(django, pin, func, instance, args, kwargs):
+    """Patches django.core.handlers.base.BaseHandler.load_middleware to patch all middlewares."""
+    settings_middleware = []
+    if hasattr(django.conf.settings, 'MIDDLEWARE') and django.conf.settings.MIDDLEWARE:
+        settings_middleware += django.conf.settings.MIDDLEWARE
+    if hasattr(django.conf.settings, 'MIDDLEWARE_CLASSES') and django.conf.settings.MIDDLEWARE_CLASSES:
+        settings_middleware += django.conf.settings.MIDDLEWARE_CLASSES
+
+    for mw_path in settings_middleware:
+        split = mw_path.split('.')
+        if len(split) > 1:
+            mod = '.'.join(split[:-1])
+            attr = split[-1]
+            mw = django.utils.module_loading.import_string(mw_path)
+            if isfunction(mw) and not isinstance(mw, wrapt.ObjectProxy):
+                # Function-based middleware is a factory which returns the function used to handle the requests.
+                # So instead of wrapping the factory, we want to wrap its returned value.
+                def traced_factory(func, instance, args, kwargs):
+                    # r is the middleware handler function returned from the factory
+                    r = func(*args, **kwargs)
+                    return FunctionWrapper(r, traced_func(django, 'django.middleware', resource=mw_path))
+                wrap(mod, attr, traced_factory)
+            elif isclass(mw):
+                for hook in ['process_request', 'process_response', 'process_view',
+                             'process_exception', 'process_template_response', '__call__']:
+                    if hasattr(mw, hook) and not isinstance(getattr(mw, hook), wrapt.ObjectProxy):
+                        wrap(mod, '{0}.{1}'.format(attr, hook),
```

Comment:
It is supposed to be `mod` but `mod` isn't a good name for what the variable represents.

All I'm doing is splitting the middleware path into two parts because `wrap` needs two arguments for the thing being patched.

I'll see if I can rename so things are more clear.

Reply:
We can actually do away with it here, but for the other one (function-based mw) we still need it.

Reply:
oh, got it, it is suppose to be the module name, not the actual module.

definitely need some clarity here.

Reply:
I made some changes here, let me know if it's more clear now 😄